src/index/src/fulltext_index/create/tantivy.rs

```diff
@@ -0,0 +1,316 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::Path;
+
+use async_trait::async_trait;
+use snafu::ResultExt;
+use tantivy::schema::{Schema, TEXT};
+use tantivy::store::{Compressor, ZstdCompressor};
+use tantivy::tokenizer::{LowerCaser, SimpleTokenizer, TextAnalyzer, TokenizerManager};
+use tantivy::{doc, Index, SingleSegmentIndexWriter};
+use tantivy_jieba::JiebaTokenizer;
+
+use crate::fulltext_index::create::FulltextIndexCreator;
+use crate::fulltext_index::error::{IOSnafu, Result, TantivySnafu};
+use crate::fulltext_index::{Analyzer, Config};
+
+pub const TEXT_FIELD_NAME: &str = "greptime_fulltext_text";
+
+/// `TantivyFulltextIndexCreator` is a fulltext index creator using tantivy.
+///
+/// Here use a single segment to store the index so the maximum capacity for
+/// the index is limited to 2<<31 rows (around 2 billion rows).
+pub struct TantivyFulltextIndexCreator {
+    /// The tantivy index writer.
+    writer: Option<SingleSegmentIndexWriter>,
+
+    /// The field for the text.
+    text_field: tantivy::schema::Field,
+}
+
+impl TantivyFulltextIndexCreator {
+    /// Creates a new `TantivyFulltextIndexCreator`.
+    ///
+    /// The `path` is the directory path in filesystem to store the index.
+    pub async fn new(path: impl AsRef<Path>, config: Config, memory_limit: usize) -> Result<Self> {
+        tokio::fs::create_dir_all(path.as_ref())
+            .await
+            .context(IOSnafu)?;
+
+        let mut schema_builder = Schema::builder();
+        let text_field = schema_builder.add_text_field(TEXT_FIELD_NAME, TEXT);
+        let schema = schema_builder.build();
+
+        let mut index = Index::create_in_dir(path, schema).context(TantivySnafu)?;
+        index.settings_mut().docstore_compression = Compressor::Zstd(ZstdCompressor::default());
+        index.set_tokenizers(Self::build_tokenizer(&config));
+
+        let memory_limit = Self::sanitize_memory_limit(memory_limit);
+
+        let writer = SingleSegmentIndexWriter::new(index, memory_limit).context(TantivySnafu)?;
+        Ok(Self {
+            writer: Some(writer),
+            text_field,
+        })
+    }
+
+    fn build_tokenizer(config: &Config) -> TokenizerManager {
+        let mut builder = match config.analyzer {
+            Analyzer::English => TextAnalyzer::builder(SimpleTokenizer::default()).dynamic(),
+            Analyzer::Chinese => TextAnalyzer::builder(JiebaTokenizer {}).dynamic(),
+        };
+
+        if !config.case_sensitive {
+            builder = builder.filter_dynamic(LowerCaser);
+        }
+
+        let tokenizer = builder.build();
+        let tokenizer_manager = TokenizerManager::new();
+        tokenizer_manager.register("default", tokenizer);
+        tokenizer_manager
+    }
+
+    fn sanitize_memory_limit(memory_limit: usize) -> usize {
+        // Port from tantivy::indexer::index_writer::{MEMORY_BUDGET_NUM_BYTES_MIN, MEMORY_BUDGET_NUM_BYTES_MAX}
```

Comment:
Can we just use those consts in tantivy crate?

Reply:
out-of-topic: Maybe we should limit the total row number of each SST file in compaction and flush.

Reply:
tanvity does not export these consts